base/strings/util.jl

```diff
@@ -464,11 +464,89 @@ function hex2bytes(s::AbstractString)
 end
 
 """
+    hex2bytes(s::AbstractVector{UInt8})
+
+Convert the hexadecimal bytes array to its binary representation. Returns
+`Vector{UInt8}`, i.e. a vector of bytes.
+"""
+@inline function hex2bytes(s::AbstractVector{UInt8})
+    len = length(s)
+    if isodd(len)
+        throw(ArgumentError("source vector length must be even"))
+    end
+    d = zeros(UInt8, div(len, 2))
+    return hex2bytes!(d, s)
+end
+
+"""
+    hex2bytes!(d::AbstractVector{UInt8}, s::AbstractVector{UInt8})
+
+Converts the hexadecimal bytes vector to its binary representation. The results are
+populated into a destination vector. The function returns the destination vector.
+
+# Examples
+```jldoctest
+julia> s = UInt8["01abEF"...]
+6-element Array{UInt8,1}:
+ 0x30
+ 0x31
+ 0x61
+ 0x62
+ 0x45
+ 0x46
+
+julia> d = zeros(UInt8, 3)
+3-element Array{UInt8,1}:
+ 0x00
+ 0x00
+ 0x00
+
+julia> hex2bytes!(d, s)
+3-element Array{UInt8,1}:
+ 0x01
+ 0xab
+ 0xef
+```
+"""
+function hex2bytes!(d::AbstractVector{UInt8}, s::AbstractVector{UInt8})
+    i, j = start(s), 0
+    # This line is important as this ensures computation happens in word boundary and not
+    # byte boundary. Boundary computation can be almost 10 times slower
+    n::UInt = 0
+    c1::UInt = 0
+    c2::UInt = 0
+    while !done(s, i)
+        n = 0
+        c1, i = next(s, i)
+        done(s, i) && throw(ArgumentError("source vector length must be even"))
+        c2, i = next(s, i)
+        n = number_from_hex(c1)
+        n <<= 4
+        n += number_from_hex(c2)
+        d[j+=1] = (n & 0xFF)
```

Comment:
The whole purpose of introducing UInt8 method was not to deteriorate performance significantly over the hex2bytes(s::AbstractString) method. And in the end it beats its performance by 10-15% at least. I guess before doing any changes to the code in terms of style and coding practices performance benchmarking be carried out. At this point I will not be able to invest further effort on this. But some of these typically compilers automatically identify and optimize but did not seem to be byte boundary computation was one of them. 

The ideal code for this method should be written by taking 4 bytes at a time (sanitize the input rather than error check the input) and optimize the code for SIMD as this is supposed to be a very low level function for IO operations.

Hence, unless ignoring this comment unless there is a supporting benchmark data.